### PR TITLE
Automatically render wiki TOC

### LINKED ISSUE #822: TOC for Wiki Pages
- Gitea version (or commit ref): 1.0.1
- Git version: 2.11.0
- Operating system: Arch Linux
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
This is an "clone" of gogs issue 3931
Having an TOC in Wiki Pages could be helpfull.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/41580547-toc-for-wiki-pages?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by zeripath**
Automatically add sidebar in the wiki view containing a TOC for the wiki page.

Fix #822

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by lunny** [comment]
I think the problem is the toc need a hide button but that could be another PR.

**Comment by zeripath** [comment]
Happy for this to be moved into 1.17 btw.

**Comment by zeripath** [comment]
> I think the problem is the toc need a hide button but that could be another PR.

done

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
> > I think the problem is the toc need a hide button but that could be another PR.
> 
> done

I would like it becomes a per-user setting to show or hide the directory defaultly and also we need a more width screen for the main area.

File: modules/templates/helper.go [hunk 1]
```diff
@@ -390,6 +390,42 @@ func NewFuncMap() []template.FuncMap {
...
+			case *uint64:
+				count = *val
+			case int64:
+				count = uint64(val)
+			case *int64:
+				count = uint64(*val)
+			case int:
+				count = uint64(val)
+			case *int:
+				count = uint64(*val)
```
Comment (Gusted) [inline modules/templates/helper.go]: This sounds unsafe, not sure which kind of values will ever end up here, but if they are negative values, you will end up with a huge count number.

**Review by Gusted — COMMENTED** [review]
Otherwise lgtm, just the helper function can cause some issues when used by future code.

**Review by Gusted — APPROVED** [review]
LGTM

**Review by wxiaoguang — APPROVED** [review]

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
Sorry for asking but I am not able to find the information in the code diff.

How should a markdown file look like that Gitea render an automatic TOC for it? Something like this?

```
[[TOC]]

# Header A
foobar

## Header A.a
Foobar

# Header B
Foobar
```

And how would the result look like? Do you have a screenshot?

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
> You can try it on try.gitea.io: https://try.gitea.io/wxiaoguang/test/wiki/test-toc

I assume I miss understand something. This link goes to a wiki page. Does this work for markdown-files (e.g. README.md) also?

I also would say it is not a good idea to render the TOC as a separate element. It is lost where it is now. A toc should be there where the TOC-Placeholder (e.g. something like `[[toc]]`) is placed.

Let me give you an example:
https://codeberg.org/buhtz/hyperorg/src/branch/main/README.md

Here you see
 1. The toc is part of the rendered README file.
 2. The toc is not on top but the third section (or second firts-level section) because I (the user) want it there.
 3. But the toc is hand written not autogenerated.

**Comment by wxiaoguang** [comment]
> Does this work for markdown-files (e.g. README.md) also?

Nope, this PR is only for Wiki TOC.

> I also would say it is not a good idea to render the TOC as a separate element. It is lost where it is now. A toc should be there where the TOC-Placeholder (e.g. something like [[toc]]) is placed.
> And for README ....

Since this PR has got approvals and got merged, there could be new issues for the features & enhancements.

**Comment by zeripath** [comment]
Add this to the top of your markdown file.

```md
---
gitea: none
include_toc: true
---
```

**Comment by Codeberg-AsGithubAlternative-buhtz** [comment]
> ```lisp
> ---
> gitea: none
> include_toc: true
> ---
> ```
Where can I find the documentation about it? The term `include_toc` give me any results on https://docs.gitea.io/en-us/search/?s=toc_include.

Are there some more modifiers for the toc? e.g. I don't like the fold-mechanism and the headline is not in h1-style.

**Pull request merged**